sdk/pinocchio/src/program.rs

```diff
@@ -166,21 +166,15 @@ pub fn slice_invoke_signed(
 
 /// Invoke a cross-program instruction but don't enforce Rust's aliasing rules.
 ///
-/// This function does not check that [`Ref`]s within [`Account`]s are properly
-/// borrowable as described in the documentation for that function. Those checks
-/// consume CPU cycles that this function avoids.
+/// This function does not check that [`Account`]s are properly borrowable.
+/// Those checks consume CPU cycles that this function avoids.
 ///
 /// # Safety
 ///
 /// If any of the writable accounts passed to the callee contain data that is
 /// borrowed within the calling program, and that data is written to by the
 /// callee, then Rust's aliasing rules will be violated and cause undefined
 /// behavior.
-///
-/// # Important
-///
-/// The accounts on the `account_infos` slice must be in the same order as the
-/// `accounts` field of the `instruction`.
```

Comment:
Why did you remove this? It's pretty important!

Reply:
It does not apply to the `*_unchecked` version since it does not validate that all accounts have been passed in, so the order is not important.

Reply:
Hold on, what does the syscall do if things are out of order? Does it succeed or fail? `invoke_signed` does the work to put things in the right order, and this does not. If the order doesn't matter, shouldn't we remove that logic in `invoke_signed`?

Reply:
It succeeds, the syscall finds the accounts by pubkey: https://github.com/anza-xyz/agave/blob/cfaf849b9385cc7f85c8086c334f3f2b893590f6/programs/bpf_loader/src/syscalls/cpi.rs#L917-L919

The order requirement only exists on pinocchio to optimize this check that the SDK does: https://github.com/anza-xyz/solana-sdk/blob/master/cpi/src/lib.rs#L257-L270, which makes multiple comparisons of pubkeys.

That is one of the main differences between the "checked" vs "unchecked" version – in the "checked" one, it validates that all accounts have been passed in and they can be borrowed.

Reply:
Ah gotcha, that makes sense, thanks!